### PR TITLE
repos: Introduce Python dependency repos integration

### PR DESCRIPTION
**Pull request by stefanhengl**
Relates to #32659

This change adds support for Python dependency search based on poetry.lock files.

## Test plan
- new unit tests
- new integration tests
- ran a local instance of Sourcegraph with a Python external service an verified that dependency search works on textualize/rich (contains a poetry.lock file)

### CODE REVIEW

File: internal/types/secret.go [hunk 1]
```diff
@@ -238,6 +227,37 @@ func (es *edits) redactURL(s string, path ...interface{}) error {
 	return nil
 }
 
+func (es *edits) unredactURLs(new, old []string) (err error) {
```
Comment (stefanhengl) [inline internal/types/secret.go:230]: Add test case for Python to `TestExternalService_UnredactConfig`

File: internal/extsvc/pypi/client.go [hunk 2]
```diff
@@ -62,24 +63,93 @@ type Client struct {
...
+	f, err := FindVersion(version, files)
+	if err != nil {
+		return File{}, errors.Wrapf(err, "project: %q", project)
+	}
+
+	return f, nil
+}
+
+// FindVersion finds the File for the given version amongst files from a project.
+func FindVersion(version string, files []File) (File, error) {
```
[the remark this answers is not in the document]
Comment (stefanhengl) [reply internal/extsvc/pypi/client.go:100]: done

File: internal/extsvc/pypi/client.go [hunk 3]
```diff
@@ -210,8 +280,83 @@ func (c *Client) Download(ctx context.Context, url string) ([]byte, error) {
...
+
+	switch ext := ExtN(filename, 2); ext {
+	case ".tar.gz", ".tar.bz2", ".tar.xz", ".tar.Z":
+		return ext
+	default:
+		return ""
+	}
+}
+
+func ExtN(path string, n int) (ext string) {
```
[the remark this answers is not in the document]
Comment (stefanhengl) [reply internal/extsvc/pypi/client.go]: done

File: internal/extsvc/pypi/client.go [hunk 4]
```diff
@@ -210,8 +280,83 @@ func (c *Client) Download(ctx context.Context, url string) ([]byte, error) {
...
+}
+
+// isSDIST returns true if filename belongs to a source distribution. If isSDIST
+// is true then the first return value is the full extension.
+//
+// Examples:
+//   request-1.12.2.tar.gz -> .tar.gz, true
+//   request-1.12.2.foo -> "", false
+//
+func isSDIST(filename string) string {
```
[the remark this answers is not in the document]
Comment (stefanhengl) [reply internal/extsvc/pypi/client.go:320]: done

File: internal/types/secret.go (hunk 1 above)
Comment (stefanhengl) [reply internal/types/secret.go:230]: done

**Review by tsenart — APPROVED** [review]

**Pull request merged**